server/svix-server/src/queue/rabbitmq.rs

```diff
@@ -0,0 +1,269 @@
+
+    Ok((producer, consumer))
+}
+
+async fn declare_delayed_message_exchange(channel: &lapin::Channel) -> Result<String> {
+    let exchange_name = "first-message";
+
+    // See https://www.rabbitmq.com/amqp-0-9-1-reference.html#exchange.declare
+    let opts = ExchangeDeclareOptions {
+        // Want the server to create the exchange if it doesn't already exist
+        passive: false,
+        nowait: false,
+        // Want the exchange to survive restarts, and not be deleted even when the underlying queues are
+        // deleted
+        durable: true,
+        auto_delete: false,
+        internal: false,
+    };
+
+    // See https://github.com/rabbitmq/rabbitmq-delayed-message-exchange#usage
+    let mut args = FieldTable::default();
+    args.insert(
+        "x-delayed-type".into(),
+        AMQPValue::LongString("direct".into()),
+    );
+
+    ctx!(
+        channel
+            .exchange_declare(
+                exchange_name,
+                lapin::ExchangeKind::Custom("x-delayed-message".into()),
+                opts,
+                args
+            )
+            .await
+    )?;
+
+    Ok(exchange_name.to_owned())
+}
+
+async fn declare_bound_queue(
+    queue_name: &str,
+    exchange_name: &str,
+    channel: &lapin::Channel,
+) -> Result<()> {
+    // Ref https://www.rabbitmq.com/amqp-0-9-1-quickref.html#queue.declare
+    let opts = QueueDeclareOptions {
+        // If the queue already exists with the same configuration, we want the server to respond w/OK
+        // Alternatively, if the queue exists with a _different_ configuration, we want this to fail
+        passive: false,
+        nowait: false,
+
+        // We want to support multiple consumers per queue
+        exclusive: false,
+
+        // We want the queue to survive rust-primary restarts
+        durable: true,
+        auto_delete: false,
+    };
+
+    // Refs https://www.rabbitmq.com/maxlength.html#definition-using-x-args and https://www.rabbitmq.com/dlx.html#using-optional-queue-arguments
+    // We may want to figure out what the queue length enforcement looks like and dead letter queueing at a later point in time
+    let args = FieldTable::default();
+    ctx!(channel.queue_declare(queue_name, opts, args).await)?;
+
+    let routing_key = queue_name;
+    ctx!(
+        channel
+            .queue_bind(
+                queue_name,
+                exchange_name,
+                routing_key,
+                QueueBindOptions { nowait: false },
+                FieldTable::default()
+            )
+            .await
+    )?;
+
+    Ok(())
+}
+
+async fn start_queue_consumer(
+    queue_name: &str,
+    channel: &lapin::Channel,
+    prefetch_size: u16,
+) -> Result<lapin::Consumer> {
+    // Ref https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.consume.consumer-tag
+    // NOTE from @svix-gabriel: it's not really clear why this needs to be set to anything,
+    // so letting rabbit choose a default for now
+    let consumer_tag = "";
```

Comment:
A common case is to encode the software version (think like a user agent), or "role" in this iirc. Useful if/when deploys go sour and maybe you've got a lingering consumer running old code. Otherwise, they all look the same in the list.

Reply:
Updated, although injected some randomness based on https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.consume.consumer-tag